### ISSUE TITLE
Add automated testing suite for Rescuezilla's long-term maintainability

### ISSUE THREAD

**Issue by shasheene** [rescuezilla/rescuezilla#151]
For all Rescuezilla releases until now, the project has been tested manually. This is very painful and tedious process:

* Creating virtual hard drives with legacy MBR (DOS) partition table and GPT partition tables
* Adding many different filesystems to these drives.
* Making backups of these drives using eg, _Redo Backup and Recovery_ v1.0.4, Rescuezilla v1.0.5 or a version of Clonezilla
* Compare the contents of a Clonezilla image to a Rescuezilla image using the graphical diff tool `meld`.
* Testing it can restore using Rescuezilla.
* Same for Windows 10 MBR, Windows 10 GPT, Linux MBR, Linux GPT (and now LVM-based operating systems like CentOS)
* Test whether the Rescuezilla backup can be restored using Clonezilla

It's worth noting that Clonezilla has been around for almost 20 years and does **not** have an automated test suite.

Given Rescuezilla [only has 18 Patreon supporters](https://www.patreon.com/join/rescuezilla), the only way for Rescuezilla to be a sustainably maintainable side project over the next 50+ years (see: [Rescuezilla roadmap](https://github.com/rescuezilla/rescuezilla/wiki/Rescuezilla-Project-Roadmap)) whilst having full backwards compatibility is to lower the long-term maintenance burden by developing a robust and fast automated testing suite. If implemented correctly an automated test suite will allow bugs to be found before release without the huge time commitment it currently needs.

To exercise all code paths effectively, perhaps this is best implemented as a Python script which executes Rescuezilla's GTK graphical element's signal handlers (to "click" the GUI and select source/destination etc). It's definitely worth evaluating if GTK testing tools like [dogtail](https://wiki.ubuntu.com/Testing/Automation/DogtailTutorial) is a better approach. It will be a lot of work to maintain this script but it means that the behavior of hundreds (and in future thousands) of historical disk images can be quickly tested.

Stretch goal: Ideally the bootability of drives restored from an operating system images would be tested, eg launch it using Qemu and once booted the operating system could open listening port on the network, which the test script could then connect to.

Side note: It makes sense for the test files to be managed in a separate Git repository to the source code. given the files (eg, Windows 10 disk image created by _Redo Backup and Recovery_ v1.0.4) will be large binary files, and GitHub has a limit of the LFS (Large File Storage) storage space and bandwidth usage.

Given the scope of this task it may be some time until it is completed.

**Comment by shasheene** [comment]
An alternative approach to "clicking" the GUI with a Python script is to implement [task #106 ("Prepopulating fields for (semi) automating a restore operation")](https://github.com/rescuezilla/rescuezilla/issues/106).

This would add YML file configuration for unattended Rescuezilla restore operations, which could be leveraged for the test suite. 

Also Clonezilla's unattended mode can be used by eg, calling ["ocs-sr" with some arguments](https://clonezilla.org/show-live-doc-content.php?topic=clonezilla-live/doc/05_Preseed_options_to_do_job_after_booting). This could be used to test Clonezilla/Rescuezilla compatibility.

**Comment by candrapersada** [comment]
Backups created with Clonezilla and later can be restored using the Rescuezilla v2.1.1 version or not?

**Comment by shasheene** [comment]
Yes, backups created with Clonezilla can be restored with Rescuezilla v2.0+. And backups created using Rescuezilla v2.0+ can be restored using Clonezilla.

Rescuezilla v2.1.1 doesn't yet exactly match Clonezilla's restore logic (task #146), but this makes no difference except in a few extremely rare cases where a [warning message is generated](https://github.com/rescuezilla/rescuezilla/issues/131#issuecomment-743733003). I will systematically deal with these extremely rare cases by implementing this GitHub task, and GitHub task #146.

**Comment by pfrouleau** [comment]
Hi @shasheene, Is this along the lines of what you are planning for this issue?
https://github.com/pfrouleau/rescuezilla/commit/99dccf5bcb2894078b2da2fd6a2f6669c08c6086

BTW, do you know `Vagrant`? It could be a way to get VMs with various OS to verify if they still boot after a restore.

**Comment by shasheene** [comment]
@pfrouleau, that's great work!

I haven't used Vangrant, but I've read a little about it before. I was considering a similar approach, but I was thinking about using QEMU rather than VirtualBox. The reason is QEMU supports both emulation and virtualization, and I'm intended on producing a build for eg, ARM systems in the future. QEMU may overlap with another issue: I am not completely sure how I am going to compile Rescuezilla for ARM systems going forward. The Docker approach used for Rescuezilla right now is only sufficient for x86_64, but not x86 or ARM because the Docker environment is running on an x86_64 system doing x86_64 compilation. As part of task #161 I was considering perhaps using the very same QEMU environment to do the build itself, in which case the build and testing could happen on the same system. That's just an idea I've been evaluating in my head.

Also as alluded to in this task, I am considering using [Dogtail to "click" the GUI](https://wiki.ubuntu.com/Testing/Automation/DogtailTutorial) to get the most realistic testing regime. This would catch errors such as when the translation file contains a mispelled variable name, because that causes an error at runtime only when the string is displayed. With a well designed test suite there's so many possible combinations of testing that could be done (eg, run the entire test suite against every supported language) with the final "report" being a single screen showing any test cases that failed, and how long they ran for.

Another consideration is the build bot should ideally run the testing suite (hopefully every release). I'll be switching away from Travis-CI.org soon (#149). Build bots often use virtualization over containerization, and virtualized system within virtualized systems may don't work -- which is why QEMU emulation may be the way to go (if performance is not terrible).

**Again these are all just random thoughts, your approach of a headless VirtualBox script would definitely be sufficient for desktop testing by developers -- which may be all that's required.** The work you've already done so far is greatly appreciated and I'll be adapting what you've done (and whatever further development you make on it).

**Comment by shasheene** [comment]
[Integration test suite](https://github.com/rescuezilla/rescuezilla/tree/master/src/integration-test) is now complete.

The automated VirtualBox workflow and boot check code complete, as are a few tests for various operating systems.

Currently it's only testing backup/restore using the Clonezilla tool within the Rescuezilla ISO image but this will be trivially expanded to include using Rescuezilla once task #258 is complete (to add a command-line interface to Rescuezilla).

Thanks @pfrouleau your initial script was useful!
